libs/model/src/services/openai/generateTokenIdea.ts

```diff
@@ -0,0 +1,175 @@
+import { blobStorage } from '@hicommonwealth/core';
+import fetch from 'node-fetch';
+import { OpenAI } from 'openai';
+import {
+  ChatCompletionMessage,
+  ChatCompletionUserMessageParam,
+} from 'openai/resources/index.mjs';
+import { v4 as uuidv4 } from 'uuid';
+import { config } from '../../config';
+import { models } from '../../database';
+import { TokenInstance } from '../../models/token';
+
+type TokenIdea = {
+  name: string;
+  symbol: string;
+  description: string;
+  imageURL: string;
+};
+
+type GenerateTokenIdeaProps = {
+  ideaPrompt?: string;
+};
+
+type GenerateTokenIdeaResponse = {
+  tokenIdea?: TokenIdea;
+  error?: string;
+};
+
+const TOKEN_AI_PROMPTS_CONFIG = {
+  name: (ideaPrompt?: string) => `
+      Please ${ideaPrompt ? `use this prompt in quotation "${ideaPrompt}" to ` : ''}
+      create a random idea for a memecoin cryptocurrency, have the token name informed by current events, 
+      and historical internet based humor. Make it hyper opinionated and subversive or topical. Provide just the 
+      token name without any acronym or symbol. Make it either cringe or amazing. Restrict your answer to between 
+      6 and 25 characters.
+    `,
+  symbol: `
+      Create a symbol for this token (symbol) of 3-6 characters. The symbol should not include special characters 
+      or emojis. Restrict your answer to between 3 and 6 characters.`,
+  description: `
+      Provide a description for the token. Make it funny and keep it in the vein of ironic, sardonic, twitter personae.
+      There are no emoji restrictions. Keep the description to 3 sentences. DO NOT BE OVERLY POSITIVE about global 
+      phenomenon, only the asset itself. Restrict your answer to between 1 and 3 sentences and less than 180 characters.
+    `,
+  image: (name: string, symbol: string) => `
+      Please create an image for a web3 token named "${name}", having a symbol of "${symbol}". 
+    `,
+};
+
+const TokenErrors = {
+  OpenAINotConfigured: 'OpenAI key not configured',
+  OpenAIInitFailed: 'OpenAI initialization failed',
+  RequestFailed: 'failed to generate complete token idea generation',
+  IdeaPromptVoilatesSecurityPolicy:
+    'provided `ideaPrompt` voilates content security policy',
+  ImageGenerationFailure: 'failed to generate image for token idea',
+};
+
+// we have to maintain this to have "conversational" context with OpenAI
+const convoHistory: (ChatCompletionMessage | ChatCompletionUserMessageParam)[] =
+  [];
+
+const chatWithOpenAI = async (prompt = '', openai: OpenAI) => {
+  convoHistory.push({ role: 'user', content: prompt }); // user msg
+
+  const response = await openai.chat.completions.create({
+    model: 'gpt-3.5-turbo',
```

Comment:
can we use newer model? like 4 turbo or 4o mini

Reply:
i mean we can but not sure if the prod chatgpt API key is setup to allow that. Will check with the product to see what our API key supports.

Reply:
ok so looks like it should, and i tried with `'gpt-4-turbo'` and it works. Updated now.